File: scala/nucleotide-count/src/main/scala/NucleotideCount.scala
class DNA(strand: String) {
  private val allowedNucleotides = Set('A', 'C', 'G', 'T')

  def nucleotideCounts: Either[Throwable, Map[Char, Int]] =
    validate.map { validStrand =>
      allowedNucleotides
        .map(c => (c, validStrand.count(_ == c)))
        .toMap
    }

  private def validate: Either[Throwable, String] = {
    if (!strand.forall(allowedNucleotides.contains)) {
      return Left(new IllegalArgumentException("Invalid input"))
    }

    Right(strand)
  }
}

File: fsharp/nucleotide-count/NucleotideCount.fs
module NucleotideCount

let nucleotideCounts (strand: string): Option<Map<char, int>> =
    let nucleotides = [ 'A'; 'C'; 'G'; 'T' ]

    let isNucleotide c = Seq.contains c nucleotides

    let countNucleotides str =
        let count n =
            str |> Seq.filter (fun c -> n = c) |> Seq.length

        nucleotides
        |> Seq.map (fun n -> n, count n)
        |> Map.ofSeq

    Some strand
    |> Option.filter (Seq.forall isNucleotide)
    |> Option.map countNucleotides
